### ISSUE TITLE
When adding servless class, preserve Class instead of going through String

### ISSUE THREAD

**Issue by jernst** [eclipse/jetty.project#894]
Line 427 of jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java reads as follows:

```
    public ServletHolder addServlet(Class<? extends Servlet> servlet,String pathSpec)
    {
        return getServletHandler().addServletWithMapping(servlet.getName(), pathSpec);
   }
```

This will determine the name of the class, and add the servlet by name. As a result, when attempting instantiate the servlet, a ClassLoader will be needed to resolve the Class. This is needlessly inefficient and risks that the context ClassLoader is different from the servlet's intended ClassLoader (which is my problem). But none of this is needed as the ServletHandler API does permit adding the class directly. So. Please change that line to
```
        return getServletHandler().addServletWithMapping(servlet, pathSpec);
```

**Comment by joakime** [comment]
Incorrect usage of Jetty.

Raw `ServletHandler` use is very very limited in functionality, its really meant for the most naive use cases.  

The `ServletHandler` class is an internal class that the `ServletContextHandler` uses to manage the servlet scope and lifecycle.

To correct your usage, use the `ServletContextHandler` as the parent context and handler, that will also set up a proper `ServletContext` for you (which raw usage of `ServletHandler` does not do).

Then use the [`ServletContextHandler.addServlet(ServletHolder holder, String pathSpec)`](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/servlet/ServletContextHandler.html#addServlet-org.eclipse.jetty.servlet.ServletHolder-java.lang.String-) method instead (if instantiation of the class is important to you).

Like this:

```java
        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        context.setBaseResource(Resource.newResource("my/base/resource"));
        server.setHandler(context);
        
        HelloServlet helloServlet = new HelloServlet();
        ServletHolder helloHolder = new ServletHolder(helloServlet);
        helloHolder.setInitParameter("greeting", "Hello ${user}");
        context.addServlet(helloHolder, "/hello/*");
```

or like this

```java
        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        context.setBaseResource(Resource.newResource("my/base/resource"));
        server.setHandler(context);

        ServletHolder helloHolder = new ServletHolder(new HelloServlet("Hello ${user}"));
        context.addServlet(helloHolder, "/hello/*");
```

If you decide to continue to use `ServletHandler`, then know as your project grows (more and more), you will eventually wind up implementing (independently) nearly every feature that `ServletContextHandler` give you out of the box, but in your own, likely non-standard way.

Signs that you are pushing the limits of raw `ServletHandler` use:

 - You are using ServletHandler.addFilter() methods
 - You have more than 1 ServletHandler
 - You have added more than 1 Servlet to the ServletHandler
 - You are having to deal with Errors and the Error Handler
 - You are wanting to use AsyncContext
 - You are wanting to use Async I/O
 - You want to use sendRedirect() or sendError() 
 - You are having to manage behavior based on DispatcherType
 - You want to use a RequestDispatcher
 - You want to include() other resources
 - You want to forward() to other resources
 - You are finding that various features of Request and/or HttpServletRequest seem to not return values
 - You want to access HttpServletRequest resources via getResource() / getResources() / getResourceAsStream() / getRealPath() 
 - You want to use JNDI
 - You want to use a 3rd party library that needs/uses the Servlet API
 - You want to access information in the ServletContext

**Comment by jernst** [comment]
Appreciate the usage comments. Using `ServletHolder` is what I'm doing already as a workaround.

However, I stand by my comment that if the `ServletContextHandler` has
```
addServlet(String servlet,String pathSpec)
addServlet(Class<? extends Servlet> servlet,String pathSpec)
```
and the `ServletHandler` has
```
addServletWithMapping(String servlet,String pathSpec)
addServletWithMapping(Class<? extends Servlet> servlet,String pathSpec)
```
that the correct way of mapping those is to preserve the Class, not converting it to String and needlessly (and perhaps incorrectly, as in my case) re-looking it up with a context ClassLoader.

**Comment by janbartel** [comment]
Suggested fix checked in to jetty-9.2.x, 9.3.x, 9.4.x and master. Thanks.
